### ISSUE TITLE
Eval loss is nan

### ISSUE THREAD

**Issue by carbonscott** [carbonscott/maxie#4]
```
05/09/2024 04:33:52 INFO __main__
[RANK 0] LOSS:EVAL - epoch 0, micro_batch 0, mean train loss = nan
05/09/2024 04:37:56 INFO __main__
[RANK 0] LOSS:EVAL - epoch 0, micro_batch 0, mean validation loss = nan
```

**Comment by carbonscott** [comment]
Use `torch.save(...)` to dump data from https://github.com/carbonscott/maxie/blob/7212038e8fe4ad7b156b6b1f5e351c5ed08413e2/train/train.fsdp.py#L500

**Comment by carbonscott** [comment]
Added data dumping statement to inspect the process:
```
@torch.no_grad()
def estimate_loss(dataloader, model, autocast_context, max_iter = None, desc = '', device = 'cpu'):
    ''' Estimate loss.
        The dataloader should be wrapped with Dataloader class or
        DistributedSampler class, best with shuffle being true.  The shuffle
        takes place before batching.
    '''
    model.eval()

    if max_iter is None:
        max_iter = len(dataloader)

    losses      = torch.zeros(len(dataloader), device = device)
    num_samples = torch.zeros(len(dataloader), device = device)
    masks       = torch.zeros(len(dataloader), device = device)
    for enum_idx, batch_data in tqdm.tqdm(enumerate(dataloader), total = max_iter, desc = f'[RANK {dist_rank}] Eval{desc}'):    # (B, C, H, W)
        # Sample at most max_iter batches
        if enum_idx >= max_iter: break

        ## print("Pre fetching")

        # Skip a batch if it's a None
        if batch_data is None: continue

        batch_input = batch_data
        batch_input = batch_input.to(device, non_blocking = True)

        ## print("Post fetching")

        with autocast_context:
            ## print("Forwarding")
            batch_output = model(batch_input)

            ## print("Loss")
            loss = batch_output.loss

        # !! Data dump
        if dist.get_rank() == 0:
            data_dump = {
                "batch_data" : batch_data,
                "batch_output" : batch_output,
                "loss" : loss,
            }
            torch.save(data_dump, f'data_dump.loop_{enum_idx:06d}.pt')

        losses[enum_idx]      = loss
        num_samples[enum_idx] = len(batch_input)
        masks[enum_idx]       = 1

    losses_sum      = torch.dot(losses[masks > 0], num_samples[masks > 0])  # [WORKAROUND] Need a proper logic to filter out None events
    num_samples_sum = num_samples[masks > 0].sum()

    world_losses_sum      = [ torch.tensor(0.0).to(device) for _ in range(dist_world_size) ]
    world_num_samples_sum = [ torch.tensor(0.0).to(device) for _ in range(dist_world_size) ]
    dist.all_gather(world_losses_sum, losses_sum)
    dist.all_gather(world_num_samples_sum, num_samples_sum)

    world_losses_mean = torch.tensor(world_losses_sum).sum() / torch.tensor(world_num_samples_sum).sum()

    # !! Data dump
    if dist.get_rank() == 0:
        data_dump = {
            "world_losses_sum" : world_losses_sum,
            "wolrd_num_samples_sum" : wolrd_num_samples_sum,
            "world_losses_mean" : world_losses_mean,
        }
        torch.save(data_dump, f'data_dump.end_{enum_idx:06d}.pt')

    model.train()

    return world_losses_mean
```

**Comment by carbonscott** [comment]
`world_losses_sum` prints 

```
[tensor(1646125.8750),
 tensor(1453168.3750),
 tensor(1273263.2500),
 tensor(nan),
 tensor(1887932.7500),
 tensor(1296456.7500),
 tensor(1640379.),
 tensor(nan),
 tensor(343279.5312),
 tensor(4948819.5000),
 tensor(3890954.),
 tensor(1033767.5000),
 tensor(685839.4375),
 tensor(2139664.7500),
 tensor(nan),
 tensor(191.8651),
 tensor(nan),
 tensor(nan),
 tensor(nan),
 tensor(1520167.),
 tensor(2387776.5000),
 tensor(1096222.5000),
...
]
```

I suspect the issue might be related to numerical overflow, but need more evidence.

**Comment by carbonscott** [comment]
At scale, the distributed eval doesn't bring a huge benefit (need to deal with calculating the true world mean loss).  Try to perform eval on rank 0 only.
